### DISCUSSION TITLE
Saving Ctrlr datas within Host session

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#291] [Q&A]
Hi,
I worked with stateData to save user banks informations and the last preset parameters before quiting the app.
It works very well but if ever I work on a project with cubase hosting my panel as a vst, the stateData valuetree will be the same for whatever session I am working on. So if I work on 2 different cubase projects the recall will be the same for both projects.
stateData is more fitted to save global panel values. It's perfect for user banks or global settings (midi i/o, paths etc)

Is there a way to save a memoryblock containing an hexstring of the program parameter values inside a Cubase or any DAW session file?
I thought about making an invisible uiLabel containing the program sysex dump  hex string, this modulator's  property "export parameter to VST host" would be set to "export", then it should be saved by the host as a VST parameter value?
On launch This would make it possible to get the string from this label and send it back to the machine this way the program is entirely saved within the host session. I am not sure about how it really works between datas stored by the host and the ctrlr plugin instance.

Is it the  proper way to go or do you have another common practice for that?

Thanks for your insight, have a nice day.

Damien

EDIT: 
1. After trying multiple ideas I came to the conclusion that this saveData function is not that much related specifically to the VST host. Sure it's a trick to save the last state of the panel and recall it, but it's just a property value of the panel so if you work on several projects You'll recall the latest state of the pannel whatever session you're in and that's it. There must be a way to recall a memoryblock stored in the plugin prefs within the host session like any other plugins. But I really don't know how it works.
2. As a VST plugin my panel doesn't send the midi messages I execute in my LuaPanelLoaded script. In Ctrlr App it's working. Is it normal?
3. As a VST plugin is the LuaPannelLoaded before or after the LuaPannelRestoreState ?

Thanks ;)

**Comment by Godlike-Productions** [comment]
I'm using JSON to save data globally and external to the plugin - https://github.com/RomanKubiak/ctrlr/discussions/196

You need to convert the MemoryBlock to a HexString to save, and convert back on loading.  I think I've outlined my methods above, but if you need more info, let me know and I'm happy to share.

Also MIDI from VST is broken in ctrlr versions on windows after 5.3.201.  I'm dealing with this at the moment as well.

**Reply by damiensellier** [reply]
Thanks! It seems very intersting when you need to store extra datas like tags for presets etc I'll probably use it when I need to improve the librarian side of my pannels.

From what I understood after reading the JUCE forum, it seems that the statedata valuetree we have also in CTRLR acts like a bridge between the plugin and the host. So when the host session is saved, (ctrl+s in cubase for ex.) it gets the plugin valuetree state datas and save them in the host file (.cpr). So yes it's stored in the ctrlr cache but it also should be saved within the host. I'll try tomorrow night to check why I cannot recall (yet) my current preset in cubase, but it should be there ready to be saved and restored. 

 

> When the host wants to store the project, it will ask your plugin to return it’s current state by calling getStateInformation(). Same when it loads a project, it will call your setStateInformation() providing the same chunk of memory allowing your plugin to restore what ever it needs to restore.
> This chunk will be included by your hosts project file as binary blob.

**Reply by Godlike-Productions** [reply]
You can manually initiate that by saving CtrlrState from the save menu.  You can also save snapshots, which I think are akin to presets in the VST. These are available in the Program Menu.

The way I've treated it is that the Snapshots essentially save the state of all the parameters with VST ID's and that you can manually add stuff to the value tree.  I haven't used them with the plugins I've developed, because the sysex dumps are the important things. I will investigate further though.

I export as sysex and json, both of which are available to the user outside of the plugin and are human readable. XML is another option (and I think what the value tree uses), but json seems to be the favoured data exchange format nowdays, hence why I ultimately chose that in my plugin.

**Comment by dobo365** [comment]
My 2c ;-)

- Im using State Data but only to save single values not entire memoryblocks. As you know, the state data is coming in the temp file of the panel. So you could save one patch but not a complete bank
- what I'm then doing is storing a variable with the full path of the last saved file and I load it at panel opening (or an Init patch if not found). Loading a patch automatically at panel opening is made optional by a switch the user can set.
- I also store full pathnames for each bank that the user loaded in the librarian so I know which bank has been loaded in which slot
- I spent quite a time earlier this year (or was it last year) to identify the order of things at startup and where to put what. In my Pro-2 panel I had the bank restore in PanelLoaded but I realized with my Prologue panel that it was not working well. I also discovered that ReadStateDate was called 2 times, once before PanelLoaded and once after. It is only after the second time that the variables are restored. This is the reason I re-ordered things (and I still need to do that in the Pro-2 panel)

Here is the order I have in my Prologue panel:
- PanelLoaded: call for different Init methods + AssignModulators + check button positions on panel. At the end I call a timer 50ms
- in Timers, my 50ms timer is directly stopped.  Meanwhile, ReadStateDate has done its job. I do some further initializations (and optionaly can load a patch or init patch) then I start another 200ms timer (time needed to have all controls processed without displaying something in my LCD)
- in Timers, I stop the 200ms timer directly and then I initialize / restore the patch banks. I also show the welcome message

I would recommend to not use snapshots. I read somewhere that it was a technique developed and used to store patches but that it has not been developed further. Maybe it works fine. I don't know.

I have no issue in all my panels to send/receive Midi messages in standalone or in VST using sendMidiMessageNow(). Simple sysex or full patches.
In VST, when using the panel 2 times on different tracks, each track keeps its patch at project reload.

Hope this will help you a bit...

**Reply by damiensellier** [reply]
Thanks @Godlike-Productions  and @dobo365 for your answers I finally have some answers and have it working properly.
Beforehand I investigated like Dominique about the order of function calls on startup.
I made a script with flags for each steps triggered with a timer. And from what I can see I can confirm LuaPannelLoaded
happens after LuaPannelRestoreState.  So from my experience LuaPannelRestoreState happened first then LuaPannelLoaded second. Not sure if LuaPannelLoaded is executed twice though like @dobo365 said but it would explain why init settings are read twice.

I cannot get the initial messages sent via LuaPannelLoaded in the plugin version. I really don't know why, but I work with the latest macos version and there's some bugs...
So I set some timers and let 1000ms before the initial messages sent from the LuaPannelRestoreState script. And it worked first try.

LuaPannelSaveState is definitely a bridge to the host to store some values in the session files.
I stored the current Program Sysex (316bytes) as a MemoryBlock and restored it on launch without any issues.

To save the MemoryBlock as an hex string in the valueTree: 

```
SaveStateData = function(--[[ ValueTree --]]stateData)
	if  MbLastUsedProgram ~= nil then
		stateData:setProperty ("LastUsedProgram", MbLastUsedProgram:toHexString(1), nil)
	end -- if
end -- function
```

To read the hex string in the valueTree and convert it to a MemoryBlock: 
```
ReadStateData = function(--[[ ValueTree --]] stateData)
	if stateData:hasProperty("LastUsedProgram") == true then
		MbLastUsedProgram = MemoryBlock()
		MbLastUsedProgram:loadFromHexString(stateData:getProperty("LastUsedProgram"))
	end -- if
end -- function
```

Et voilà! Different sessions either from a VST host or from Ctrlr App don't intefere and I can have several projects running at the same time and the values are restored properly without being messed up.

Thanks a lot for your help to both of you. 

Damien

**Reply by Godlike-Productions** [reply]
Donimique,

Do you know the order that things are executed regarding:

panelLoaded
panelBeforeLoad (before Panel Loaded I assume, but where in relation to the others?)
luaPanelResourcesLoaded
luaPanelRestoreState (when is this automatically called the first time)

**Reply by damiensellier** [reply]
Hi @dobo365 I struggled with the repetition of "ReadStateData "  today, which was driving me crazy while recalling the last used preset from Cubase . After 4 hours of brainstorming I figured out a button was connected to the method and, since on startup every single button's value is reset, it triggers all the method "when value changes". So I added
`if source ==  4 then ...`  and it blocked the script playing twice.  (from the buttton value reset, and the global panel properties).
 I cannot get the ReadStateDate  twice now.  Filtering the source of the method is really important.
To come back to our observation, no, the "ReadStateData" is not called twice natively.
